JumpscaleLibs/sal/zosv2/Zosv2.py

```diff
@@ -0,0 +1,197 @@
+                    if ip.is_global:
+                        self.config.public_ip = str(ip)
+                        return (node, str(ip))
+        if version==4:
+            for node in filter(self.is_up, nodes):
+                ip = ipaddress.ip_interface(node.public_config.ipv4).ip
+                if ip.is_global:
+                    self.config.public_ip = str(ip)
+                    return (node, str(ip))
+            # need to check in ifaces
+    def find_free_wg_port(self, node):
+        ports = set(list(range(6000, 9000)))
+        used = set(node.wg_ports)
+        free = ports - used
+        return free.pop()
+
+    def new_reservation(self, user):
+        reservation_model = j.data.schema.get_from_url("tfgrid.workloads.reservation.1")
+        reservation = reservation_model.new()
+        reservation.customer_tid = user.id
+        reservation.data_reservation.expiration_provisioning = j.data.time.epoch + (60 * 10)  # 10 minutes
+        reservation.data_reservation.expiration_reservation = j.data.time.epoch + (60 * 10)  # 10 minutes
+        return  reservation
+
+    def get_user(self,name,email):
+        # should return error if doesn't exist
+        return self.explorer.actors_all.phonebook.get(name=name, email=email)
+
+    def check_ip_in_network(self, ip , network):
+        return netaddr.IPAddress(ip) in netaddr.IPNetwork(network)
+
+    def key_pair_get(self):
+        wg_private = public.PrivateKey.generate()
+        wg_public = wg_private.public_key
+
+        wg_private_base64 = wg_private.encode(Base64Encoder)
+        wg_public_base64 = wg_public.encode(Base64Encoder)
+
+        return wg_private_base64, wg_public_base64
+
+    def create_network(self,reservation,ip_range,node_id):
+        network = reservation.data_reservation.networks.new()
+        network.node_id = node_id
+        network.workload_id = 1
+        network.name = j.data.idgenerator.generateXCharID(16)
+        network.iprange = ip_range
+        return reservation , network
+
+    def access_node(self, network, ip_range, node):
+        network_range = netaddr.IPNetwork(ip_range).ip
+        # add 256 to plus 1 in network from 172.20.0.0 -> 172.20.1.0
+        network_range += 256
+        network1 = str(network_range) + "/24"
+        _, wg_private_encrypted, wg_public = j.tools.wireguard.generate_zos_keys(node.public_key_hex)
+        wg_port = self.find_free_wg_port(node)
+
+        nr1 = network.network_resources.new()
+
+        nr1.iprange = network1
+        nr1.node_id = node.node_id
+        nr1.wireguard_listen_port = wg_port
+        nr1.wireguard_public_key = wg_public
+        nr1.wireguard_private_key_encrypted = wg_private_encrypted
+
+        # add 256 to plus 1 in network from 172.20.0.0 -> 172.20.1.0
+        network_range += 256
+        network2 = str(network_range) + "/24"
+        ip = netaddr.IPAddress(network_range)
+        allowed_ip = ip.words
+
+        private_key, public_key = self.key_pair_get()
```

Comment:
It should be left the possiblity to the user to specify it's own wireguard public key if he already has one.
Generate a new key pair only if it is required.